Taking the real part of an expression leads to failed compilation.

I am trying to compute the squared modulus of a complex tensor. The following code fails to compile.
```cpp
#include <iostream>
#include <complex>
#include <xtensor/xtensor.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xnorm.hpp>
#include <xtensor/xcomplex.hpp>

using cpx = std::complex<double>;
using std::complex;

template<class T>
xt::xtensor<T, 2> modulus_sq(xt::xtensor<complex<T>, 2> &a)
{
  return xt::real(a * xt::conj(a));
}

int main()
{
  xt::xtensor<cpx, 2> a = {
    { cpx(1, 1), cpx(-1, 1), cpx(-2, -2) },
    { cpx(-1, 0), cpx(0, 1), cpx(2, 2) }
  };

  // std::cout << a << std::endl;
  std::cout <<  modulus_sq(a) << std::endl;

}
```
The compilation error message is the following.
```
g++ -I /Users/scheibler/anaconda3/envs/mixiva/include ./test.cpp -o test_norm -std=c++14
In file included from ./test.cpp:4:
In file included from /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xtensor.hpp:18:
In file included from /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xbuffer_adaptor.hpp:20:
In file included from /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xstorage.hpp:21:
In file included from /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xtensor_simd.hpp:14:
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xutils.hpp:1192:31: error: implicit instantiation of undefined template
      'xt::rebind_container<long, xt::xfunction<xt::detail::multiplies, const xt::xtensor_container<xt::uvector<std::__1::complex<double>,
      std::__1::allocator<std::__1::complex<double> > >, 2, xt::layout_type::row_major, xt::xtensor_expression_tag> &,
      xt::xfunction<xt::math::conj_impl_fun, const xt::xtensor_container<xt::uvector<std::__1::complex<double>,
      std::__1::allocator<std::__1::complex<double> > >, 2, xt::layout_type::row_major, xt::xtensor_expression_tag> &> > >'
        using type = typename rebind_container<std::ptrdiff_t, S>::type;
                              ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtl/xmeta_utils.hpp:82:35: note: in instantiation of template class
      'xt::get_strides_type<xt::xfunction<xt::detail::multiplies, const xt::xtensor_container<xt::uvector<std::__1::complex<double>,
      std::__1::allocator<std::__1::complex<double> > >, 2, xt::layout_type::row_major, xt::xtensor_expression_tag> &,
      xt::xfunction<xt::math::conj_impl_fun, const xt::xtensor_container<xt::uvector<std::__1::complex<double>,
      std::__1::allocator<std::__1::complex<double> > >, 2, xt::layout_type::row_major, xt::xtensor_expression_tag> &> > >' requested here
            using type = typename F::type;
                                  ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtl/xmeta_utils.hpp:86:26: note: in instantiation of template class
      'xtl::mpl::eval_if_c<false, xt::detail::expr_backstrides_type<xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > >, xt::get_strides_type<xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > > >' requested here
        struct eval_if : eval_if_c<B::value, T, F>
                         ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtl/xmeta_utils.hpp:91:9: note: in instantiation of template class
      'xtl::mpl::eval_if<xt::has_strides<xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> >, void>,
      xt::detail::expr_backstrides_type<xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > >, xt::get_strides_type<xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > > >' requested here
        using eval_if_t = typename eval_if<B, T, F>::type;
        ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xfunctor_view.hpp:128:44: note: in instantiation of template type alias 'eval_if_t'
      requested here
        using backstrides_type = xtl::mpl::eval_if_t<has_strides<xexpression_type>,
                                           ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xfunctor_view.hpp:372:34: note: in instantiation of template class
      'xt::xfunctor_applier_base<xt::detail::offset_forwarder<double, 0>, xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > &>' requested here
    class xfunctor_view : public xfunctor_applier_base<F, CT>,
                                 ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xcomplex.hpp:46:24: note: in instantiation of template class
      'xt::xfunctor_view<xt::detail::offset_forwarder<double, 0>, xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > &>' requested here
                return xoffset_view<xclosure_t<E>, real_type, 0>(std::forward<E>(e));
                       ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xcomplex.hpp:79:110: note: in instantiation of function template specialization
      'xt::detail::complex_helper<true>::real<xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > &>' requested here
                return detail::complex_helper<xtl::is_complex<typename std::decay_t<E>::value_type>::value>::real(e);
                                                                                                             ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xcomplex.hpp:117:91: note: in instantiation of function template specialization
      'xt::detail::complex_expression_helper<true>::real<xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > >' requested here
        return detail::complex_expression_helper<is_xexpression<std::decay_t<E>>::value>::real(std::forward<E>(e));
                                                                                          ^
./test.cpp:15:14: note: in instantiation of function template specialization 'xt::real<xt::xfunction<xt::detail::multiplies, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, xt::xfunction<xt::math::conj_impl_fun, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &> > >' requested here
  return xt::real(a * xt::conj(a));
             ^
./test.cpp:26:17: note: in instantiation of function template specialization 'mynorm<double>' requested here
  std::cout <<  mynorm(a) << std::endl;
                ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xutils.hpp:80:12: note: template is declared here
    struct rebind_container;
           ^
1 error generated.
```
I am using Apple LLVM on Mojave (10.14.4).
```
g++ --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/c++/4.2.1
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -28,7 +28,7 @@ install:
   - conda update -q conda
   - conda info -a
   - conda install gtest=1.8.0 cmake -c conda-forge
-  - conda install xtl==0.6.2 xsimd nlohmann_json -c conda-forge
+  - conda install xtl==0.6.4 xsimd nlohmann_json -c conda-forge
   - ps: if($env:APPVEYOR_BUILD_WORKER_IMAGE -eq "Visual Studio 2015") {
         cmake -G "NMake Makefiles" -DCMAKE_INSTALL_PREFIX=%MINICONDA%\\LIBRARY -DBUILD_TESTS=ON -DXTENSOR_USE_XSIMD=ON .
         }

--- a/.travis.yml
+++ b/.travis.yml
@@ -143,7 +143,7 @@ install:
     - conda config --set always_yes yes --set changeps1 no
     - conda update -q conda
     - conda install cmake -c conda-forge
-    - conda install xtl==0.6.2 xsimd nlohmann_json -c conda-forge
+    - conda install xtl==0.6.4 xsimd nlohmann_json -c conda-forge
     - if [[ "$ENABLE_TBB" == 1 ]]; then
         conda install tbb-devel -c conda-forge;
       fi

--- a/include/xtensor/xbuilder.hpp
+++ b/include/xtensor/xbuilder.hpp
@@ -208,6 +208,33 @@ namespace xt
         template <class T, class S>
         using get_mult_type = typename get_mult_type_impl<T, S>::type;
 
+        // These methods should be private methods of arange_generator, however thi leads
+        // to ICE on VS2015
+        template <class R, class E, class U, class X, XTL_REQUIRES(std::is_integral<X>)>
+        inline void arange_assign_to(xexpression<E>& e, U start, X step) noexcept
+        {
+            auto& de = e.derived_cast();
+            U value = start;
+
+            for (auto&& el : de.storage())
+            {
+                el = static_cast<R>(value);
+                value += step;
+            }
+        }
+
+        template <class R, class E, class U, class X, XTL_REQUIRES(xtl::negation<std::is_integral<X>>)>
+        inline void arange_assign_to(xexpression<E>& e, U start, X step) noexcept
+        {
+            auto& buf = e.derived_cast().storage();
+            using size_type = decltype(buf.size());
+            using mult_type = get_mult_type<U, X>;
+            for(size_type i = 0; i < buf.size(); ++i)
+            {
+                buf[i] = static_cast<R>(start + step * mult_type(i));
+            }
+        }
+
         template <class T, class R = T, class S = T>
         class arange_generator
         {
@@ -238,7 +265,7 @@ namespace xt
             template <class E>
             inline void assign_to(xexpression<E>& e) const noexcept
             {
-                assign_to_impl(e, m_start, m_step);
+                arange_assign_to<R>(e, m_start, m_step);
             }
 
         private:
@@ -257,30 +284,6 @@ namespace xt
             inline R access_impl() const
             {
                 return static_cast<R>(m_start);
-            }
-
-            template <class E, class U, class X, XTL_REQUIRES(std::is_integral<X>)>
-            inline void assign_to_impl(xexpression<E>& e, U start, X step) const noexcept
-            {
-                auto& de = e.derived_cast();
-                T value = start;
-
-                for (auto&& el : de.storage())
-                {
-                    el = static_cast<R>(value);
-                    value += step;
-                }
-            }
-
-            template <class E, class U, class X, XTL_REQUIRES(xtl::negation<std::is_integral<X>>)>
-            inline void assign_to_impl(xexpression<E>& e, U /*start*/, X /*step*/) const noexcept
-            {
-                auto& buf = e.derived_cast().storage();
-                using size_type = decltype(buf.size());
-                for(size_type i = 0; i < buf.size(); ++i)
-                {
-                    buf[i] = access_impl(i);
-                }
             }
         };
 

--- a/include/xtensor/xcomplex.hpp
+++ b/include/xtensor/xcomplex.hpp
@@ -60,7 +60,7 @@ namespace xt
             template <class E>
             static inline decltype(auto) real(E&& e) noexcept
             {
-                return e;
+                return std::forward<E>(e);
             }
 
             template <class E>
@@ -74,15 +74,15 @@ namespace xt
         struct complex_expression_helper
         {
             template <class E>
-            static inline auto real(E&& e) noexcept
+            static inline decltype(auto) real(E&& e) noexcept
             {
-                return detail::complex_helper<xtl::is_complex<typename std::decay_t<E>::value_type>::value>::real(e);
+                return detail::complex_helper<xtl::is_complex<typename std::decay_t<E>::value_type>::value>::real(std::forward<E>(e));
             }
 
             template <class E>
-            static inline auto imag(E&& e) noexcept
+            static inline decltype(auto) imag(E&& e) noexcept
             {
-                return detail::complex_helper<xtl::is_complex<typename std::decay_t<E>::value_type>::value>::imag(e);
+                return detail::complex_helper<xtl::is_complex<typename std::decay_t<E>::value_type>::value>::imag(std::forward<E>(e));
             }
         };
 

--- a/include/xtensor/xfunctor_view.hpp
+++ b/include/xtensor/xfunctor_view.hpp
@@ -93,7 +93,7 @@ namespace xt
                                                  get_strides_type<shape_type>>;
         using backstrides_type = xtl::mpl::eval_if_t<has_strides<xexpression_type>,
                                                      detail::expr_backstrides_type<xexpression_type>,
-                                                     get_strides_type<xexpression_type>>;
+                                                     get_strides_type<shape_type>>;
 
         using inner_shape_type = typename xexpression_type::inner_shape_type;
         using inner_strides_type = xtl::mpl::eval_if_t<has_strides<xexpression_type>,
@@ -347,8 +347,8 @@ namespace xt
         using xexpression_type = std::decay_t<CT>;
         using undecay_expression = CT;
         using functor_type = std::decay_t<F>;
-        using reference = typename functor_type::reference;
-        using const_reference = typename functor_type::const_reference;
+        using reference = decltype(std::declval<F>()(std::declval<xexpression_type>()()));
+        using const_reference = decltype(std::declval<F>()(std::declval<const xexpression_type>()()));
         using size_type = typename xexpression_type::size_type;
         using temporary_type = typename xfunctor_view_temporary_type<F, xexpression_type>::type;
     };
@@ -482,28 +482,38 @@ namespace xt
      * xfunctor_iterator declaration *
      *********************************/
 
-    template <class DT>
+    template <class R>
     struct xproxy_inner_types
     {
-        using proxy = xtl::xproxy_wrapper<DT>;
-        using pointer = typename proxy::pointer;
-        using reference = typename proxy::reference;
+        using reference = R;
+        using pointer = std::add_pointer_t<std::remove_reference_t<R>>;
     };
+
+    namespace detail
+    {
+        template <class F, class IT>
+        struct xfunctor_invoker
+        {
+            using type = decltype(std::declval<F>()(*(std::declval<IT>())));
+        };
+
+        template <class F, class IT>
+        using xfunctor_invoker_t = typename xfunctor_invoker<F, IT>::type;
+    }
 
     template <class F, class IT>
     class xfunctor_iterator : public xtl::xrandom_access_iterator_base<xfunctor_iterator<F, IT>,
                                                                        typename std::decay_t<F>::value_type,
                                                                        typename std::iterator_traits<IT>::difference_type,
-                                                                       typename xproxy_inner_types<decltype(std::declval<F>()(*(IT())))>::pointer,
-                                                                       typename xproxy_inner_types<decltype(std::declval<F>()(*(IT())))>::reference>
+                                                                       typename xproxy_inner_types<detail::xfunctor_invoker_t<F, IT>>::pointer,
+                                                                       typename xproxy_inner_types<detail::xfunctor_invoker_t<F, IT>>::reference>
     {
     public:
 
         using functor_type = F;
         using subiterator_traits = std::iterator_traits<IT>;
 
-        using proxy_inner = xproxy_inner_types<decltype(std::declval<F>()(*(IT())))>;
-        using proxy = typename proxy_inner::proxy;
+        using proxy_inner = xproxy_inner_types<detail::xfunctor_invoker_t<F, IT>>;
         using value_type = typename functor_type::value_type;
         using reference = typename proxy_inner::reference;
         using pointer = typename proxy_inner::pointer;
@@ -553,8 +563,7 @@ namespace xt
 
         using functor_type = F;
 
-        using proxy_inner = xproxy_inner_types<decltype(std::declval<F>()(*std::declval<ST>()))>;
-        using proxy = typename proxy_inner::proxy;
+        using proxy_inner = xproxy_inner_types<detail::xfunctor_invoker_t<F, ST>>;
         using value_type = typename functor_type::value_type;
         using reference = typename proxy_inner::reference;
         using pointer = std::remove_reference_t<reference>*;

--- a/include/xtensor/xoffset_view.hpp
+++ b/include/xtensor/xoffset_view.hpp
@@ -34,7 +34,7 @@ namespace xt
             template <class T>
             decltype(auto) operator()(T&& t) const
             {
-                return xtl::forward_offset<M, I>(t);
+                return xtl::forward_offset<M, I>(std::forward<T>(t));
             }
 
             template <class align, class requested_type, std::size_t N, class E,

--- a/test/test_xcomplex.cpp
+++ b/test/test_xcomplex.cpp
@@ -256,4 +256,16 @@ namespace xt
         EXPECT_TRUE(isclose(c_t(5, 5), c_t(5, -5))() == false);
 
     }
+
+    TEST(xcomplex, real_expression)
+    {
+        using cpx = std::complex<double>;
+        xtensor<cpx, 2> a = {{ cpx(1, 1), cpx(-1, 1), cpx(-2, -2) },
+                             { cpx(-1, 0), cpx(0, 1), cpx(2, 2) }};
+
+        xtensor<double, 2> exp = {{2, -2, -4},
+                                  {-2, 0, 4}};
+        xtensor<double, 2> res = real(a + a);
+        EXPECT_EQ(res, exp);
+    }
 }

--- a/test/test_xoptional.cpp
+++ b/test/test_xoptional.cpp
@@ -294,6 +294,12 @@ namespace xt
 
         reference operator()(point& p) const noexcept { return p.x(); }
         const_reference operator()(const point& p) const noexcept { return p.x(); }
+
+        // Compilation trick: a functor view on xarray_optional<point> expects the functor
+        // to provide operator() accepting reference and const_reference from xarray_optional<point>
+        // Since we never use these overloads, no need to declare the right one, nor to provide
+        // any implementation.
+        template <class T> T&& operator()(T&&) const noexcept;
     };
 
     TEST(xoptional, functor_view)

